keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,1549 @@
+ResourceClaim by the driver, a ResourceClaim can be used exclusively
+by one pod at a time, by a certain maximum number of pods, or an
+unlimited number of pods.
+
+### User Stories (Optional)
+
+#### Cluster add-on development
+
+As a hardware vendor, I want to make my hardware available also to applications
+that run in a container under Kubernetes. I want to make it easy for a cluster
+administrator to configure a cluster where some nodes have this hardware.
+
+I develop two components, one that runs as part of the Kubernetes control plane
+and one that runs on each node, and package those inside container images. YAML
+files describe how to deploy my software on a Kubernetes cluster that supports
+dynamic resource allocation.
+
+Documentation for administrators explains how the nodes need to be set
+up. Documentation for users explains which parameters control the behavior of
+my hardware and how to use it inside a container.
+
+#### Cluster configuration
+
+As a cluster administrator, I want to make GPUs from vendor ACME available to users
+of that cluster. I prepare the nodes and deploy the vendor's components with
+`kubectl create`.
+
+I create a ResourceClass for the hardware with parameters that only I as the
+administrator are allowed to choose, like for example running a command with
+root privileges that does some cluster-specific initialization for each allocation:
+```
+apiVersion: cdi.k8s.io/v1alpha1
+metadata:
+  name: acme-gpu
+driverName: gpu.acme.com
+parameters:
+  initCommand:
+  - /usr/local/bin/acme-gpu-init
+  - --cluster
+  - my-cluster
+```
+
+#### Partial GPU allocation
+
+As a user, I want to use a GPU as accelerator, but don't need exclusive access
+to that GPU. Running my workload with just 2Gb of memory is sufficient. This is
+supported by the ACME GPU hardware. I know that the administrator has created
+an "acme-gpu" ResourceClass.
+
+For a simple trial, I create a Pod directly where two containers share the same subset
+of the GPU:
+```
+apiVersion: v1
+kind: Pod
+metadata:
+  name: device-consumer
+spec:
+  containers:
+  - name: workload
+    image: my-app
+    command: ["/bin/program"]
+    podResources:
+    - name: gpu
+    resources:
+      requests:
+        memory: "64Mi"
+        cpu: "250m"
+      limits:
+        memory: "128Mi"
+        cpu: "500m"
+  - name: monitor
+    image: my-app
+    command: ["/bin/other-program"]
+    podResources:
+    - name: gpu
+    resources:
+      requests:
+        memory: "32Mi"
+        cpu: "25m"
+      limits:
+        memory: "64Mi"
+        cpu: "50m"
+  resources:
+  - name: gpu
+    template:
+      resourceClassName: "acme-gpu"
+      parameters:
+        memory: "2Gi"
+```
+
+This request triggers resource allocation on a node that has a GPU device with
+2Gi of memory available and then the Pod runs on that node. The remaining
+capacity of the GPU can be used by other pods. The resource is deallocated
+automatically once the Pod is deleted.
```

Comment:
"once the Pod is deleted because the garbage collector will also delete the ResourceClaim".

I didn't want to go into too much details here yet. Should I add that?

Reply:
I took out the slightly misleading "once the Pod is deleted". Hopefully this is better:
> The lifecycle of the resource allocation is tied to the lifecycle of the Pod.